### ISSUE TITLE
Returning a RedirectResponse in a dependency

### ISSUE THREAD

**Issue by philmade** [tiangolo/fastapi#1039]
### Description

How can I return a Redirect in a dependency to be executed in a conditional? In the example below, the code works, but I just get an exception. I want a redirect.:

```
def logged_in(user: models.User = Depends(current_user)):
    if not user.is_authenticated:
        raise HTTPException(status_code=303, detail="You need to login")

@users.get('/channel/{username}', dependencies=[Depends(logged_in)])
def channel(*, username: str,
            request: Request,
            current_user: models.User = Depends(current_user),
            db: Session = Depends(get_db))
    do_something()
```

Below is what I want, but when I use this dependency nothing happens because doesn't use the return value. I can't think of a way to do this without referencing the Dependency directly.

```
def logged_in(user: models.User = Depends(current_user)):
    if not user.is_authenticated:
        return RedirectResponse(url='/login')

@users.get('/channel/{username}', dependencies=[Depends(logged_in)])
def channel(*, username: str,
             request: Request,
            current_user: models.User = Depends(current_user),
            db: Session = Depends(get_db))
    do_something()
```

This works, but I have to put it in every single endpoint:
@users.get('/channel/{username}', dependencies=[Depends(logged_in)])

```
def channel(*, username: str,
             request: Request,
            current_user: models.User = Depends(current_user),
            db: Session = Depends(get_db)):
     
     if not user.is_authenticated:
            return RedirectResponse(url='/login')
```

**Comment by NomeChomsky** [comment]
A gut on Gitter helped me with this. The trick is to not return the Redirect - just call it. Like this:
```
def logged_in(user: models.User = Depends(current_user)):
    if not user.is_authenticated:
        RedirectResponse(url='/login')
```

**Comment by phy25** [comment]
I doubt if this works at all. At least the following does not work for me.

```python
from fastapi import FastAPI, Depends
from starlette.responses import RedirectResponse

app = FastAPI()

def redirect() -> bool:
    RedirectResponse(url='/login')
    return True

@app.get('/')
def index(should_redirect: bool = Depends(redirect)):
    return should_redirect
```

What I did before is to use custom exception handler:

```python
from fastapi import FastAPI, Depends
from starlette.requests import Request
from starlette.responses import RedirectResponse, Response

app = FastAPI()

class RequiresLoginException(Exception):
    pass

async def redirect() -> bool:
    raise RequiresLoginException

@app.exception_handler(RequiresLoginException)
async def exception_handler(request: Request, exc: RequiresLoginException) -> Response:
    return RedirectResponse(url='/login')

@app.get('/')
async def index(should_redirect: bool = Depends(redirect)) -> bool:
    return should_redirect
```

**Comment by NomeChomsky** [comment]
Yeah you're correct - I had thought that it worked but it does not. Without specifically referencing the dependency in the endpoint, I can't easily get a redirect to happen at the dependency level, it only will happen at the endpoint level.

Anyone got an idea on how to force a redirect?

**Comment by phy25** [comment]
@NomeChomsky does my exception-based solution work for you?

**Comment by NomeChomsky** [comment]
Hey I just tried that. It does work. I didn't fully understand it - so the Exception is somehow allowing a return statement to be triggered, but an *actual* return statement won't get triggered?

Its a good workaround. Would be good to be able to return / exectute a Redirect without a custom exception.

thank you for your help!

:)

**Comment by tiangolo** [comment]
Thanks for the help here @phy25 ! :cake: :clap: 

@NomeChomsky yeah, just what @phy25 said :point_up:

**Comment by bsatts** [comment]
Is there a way to define exception handlers on routers? There's currently no way to @router.exception_handler

**Comment by tiangolo** [comment]
Yeah, that's currently not supported @bsatts, only in the main app.

**Comment by eddyizm** [comment]
@NomeChomsky how did you implement that example to your user authentication? I am having the same issue but not sure how that got sorted out.

**Comment by fboerman** [comment]
@NomeChomsky @eddyizm I would also like to know how I can implement this. I want my dependency to trigger a redirect when it cant find a certain cookie. I cant find in the docs a proper way to implement this without also thouching the main app source code.

**Comment by gavrilovne** [comment]
> @NomeChomsky @eddyizm I would also like to know how I can implement this. I want my dependency to trigger a redirect when it cant find a certain cookie. I cant find in the docs a proper way to implement this without also thouching the main app source code.
> 
> @phy25 solution is therefor not possible for me

Hi!
Try it
`
async def check_auth( token: bool = Depends(check_token)):
    if not token:
        raise HTTPException(
            status_code=status.HTTP_307_TEMPORARY_REDIRECT,
            headers={'Location': '/login'})
`

**Comment by fboerman** [comment]
> > @NomeChomsky @eddyizm I would also like to know how I can implement this. I want my dependency to trigger a redirect when it cant find a certain cookie. I cant find in the docs a proper way to implement this without also thouching the main app source code.
> > @phy25 solution is therefor not possible for me
> 
> Hi! Try it
> 
> ```python
> oauth2_scheme = OAuth2PasswordBearer(tokenUrl="token", auto_error=False)
> 
> 
> async def check_auth(token: str = Depends(oauth2_scheme)):
>     if not token:
>         raise HTTPException(
>             status_code=status.HTTP_307_TEMPORARY_REDIRECT,
>             headers={'Location': '/login'})
>     return True
> 
> 
> @router.get("/login")
> async def login(request: Request):
>     return auth_template.TemplateResponse("login.html", {"request": request})
> 
> 
> @router.get("/test")
> async def test(auth: bool = Depends(check_auth)):
>     return "OK"
> ```
> 
> ![screen](https://user-images.githubusercontent.com/99733350/154408134-43f50c46-8229-403b-9e9b-f29955550055.PNG)

thanks @gavrilovne ! Thanks works perfectly!
